packages/matter.js/InterfaceNotes.ts

```diff
@@ -0,0 +1,121 @@
+// We agreed that server interface is the priority...  Here's a more detailed
+// proposal on the API.
+//
+// Note 1: This is a conceptual interface that models what the current PR is
+// generating (or will generate) for a single cluster.  I pulled it together
+// into a single place for discussion purposes.
+//
+// Question 1. We just call this "Device" but it's really a "ServerDevice".
+// In a controller we could potentially also have a "ClientDevice" that looks
+// very similar but proxies over the fabric to a remote endpoint.  So should
+// we consider making space in the API for a future Client/Server bifurcation
+// or am I thinking about this wrong?
```

Comment:
Right, so my thinking was that developers would extend Device to implement server clusters and would see a separate device for bound clusters.  I know matter sticks them both on the same endpoint but I was thinking it'd be confusing if my light switch device implementation talks to itself like it's a light bulb. 🤔

Anyway makes sense they'd share a base class.  I only brought it up now because with the approach above we might want to make separate subclasses for local vs bound/controlled devices.

Reply:
> separate subclasses for local vs bound/controlled devices.

You will know that only on runtime, so separate classes might not be possible

Reply:
On the other hand I thought about the idea of "devs extend generated classes to implement logic" ... When thinking about bridge builder I do not think that this is realistic because in fact:
You have an e.g, zwave devce, so based on it's featurees you search fr the correct of our classes ... but these can be maaaany ... so maybe it is better to always hae a handler callback approach?

Reply:
Hmm right, I've been keeping that use case in back of mind but haven't focused too much on it.  Note that there's not (I don't think) any functional difference between creating a JS class and creating an instance with handlers.  So the current stuff would work fine...  But we'd maybe want a couple of utility methods to make usage more natural.

Reply:
PS: With explicit handler registration it also migth be more easy to detect if an "own handler" was implemented ... or how can we get that info else?

Reply:
Would maybe depend on context...  Are you thinking to populate FeatureMap during discovery or to determine whether to use default implementation?  Or other use cases?

Reply:
Wether to use default implementation or a custom command handler.

Reply:
OK.  So I'm actually wiring ClusterServer in right now.  Not done yet but this is roughly where I'm headed.  WDYT?

- For each (server) cluster:
    - Search cluster exports for default handlers
    - For each attribute:
        - Create `onAttributeChange` in the device class that does nothing by default
        - Select an initial value:
            - Search `state.attrName` in user-supplied initial state
            - Fall back to TLV schema for default value
        - If there's a `getAttribute` in default handlers, add it to handler list
    - For each command:
        - Create `onCommand` in the derived class
            - If we found a handler above, `onCommand` calls that
            - If not, `onCommand` throws `Error("Not implemented")`
        - Add handler `CommandServer` that invokes `onCommand`
    - For each event:
        - Create `triggerEvent` in the derived class
            - `triggerEvent` has a TODO and throws `Error("Not implemented")`
    - Create a `ClusterServer`
        - `clusterDef` = the Cluster
        - `attributeInitialValues` = defaults collected above
        - `handlers` = handlers created above

So to answer your question, default command handlers are called unless the user overrides `onCommand` and doesn't call `super()`.  There's not a way to override attribute getters, I'm not sure that's necessary.

Reply:
Re onAttributeChange
ok, and constructire adds these methods then automatically as listeners? thats ok. Then user can still add another one OR overwrite

Re onCommand:
> If we found a handler above, onCommand calls that
it needs to work differently. handlers are used deeply internal where devcie classes are not known. Unless we want to change that (what I would not think is a good idea). In my eyes the default handlers need to know if there is a "more high level provided implementation" and then not execut his logic but the other. Thats why I did that "namedHandler approach and idea was top simply (other then now) pass the named handler into the "getDefaultHandlers" methods because then we would have the logic here like

`handler.hashandlerFor("identify") -> call that, else do my onwn minimalistic stuff.

Or how you planned to wire it that the "handlers existing on ClusterServer" (and those are used internally to handle commands) know when what to do?

So the more I think the more I would go with the "addCommandHandler("name", () func)" approach) to "Override command handlers" and then auto generate the allowed name to methiod signature list for that named handler

Re: triggerEvent:
ok, makes sense. will wire then later to a eventServer - already have it in my mind :-)

Re: "Create a clusterServer"
Here we need to think. I planned it now with this "createDefaultClusterrServer" to have anything for a meaning default - and in fact users can override defaults. I would love to stay with this.

Reply:
Re "initial attribute value": simly get it from attribute.get() ?

Reply:
Re commands: AHhhh ... read below (wireCLusterServer)  first ....

Reply:
Right sorry I threw that out quickly but I know I'm kind of violating encapsulation with how I did handlers.  I should have added more explanation.  I'll respond to subsequent comments.

Reply:
Going to resolve this conversation as it's split out in separate comments around the code now which is nice.